src/server/debugcmd.cc

```diff
@@ -68,32 +69,97 @@ struct ValueCompressInfo {
   size_t compressed_size = 0;
 };
 
-void DoPopulateBatch(std::string_view prefix, size_t val_size, bool random_value_str,
-                     const SetCmd::SetParams& params, const PopulateBatch& batch) {
-  DbContext db_cntx{batch.dbid, 0};
-  OpArgs op_args(EngineShard::tlocal(), 0, db_cntx);
-  SetCmd sg(op_args, false);
+std::string GenerateValue(size_t val_size, bool random_value, absl::InsecureBitGen* gen) {
+  if (random_value) {
+    return GetRandomHex(*gen, val_size);
+  } else {
+    return string(val_size, 'x');
+  }
+}
+
+tuple<const CommandId*, absl::InlinedVector<string, 5>> GeneratePopulateCommand(
+    string_view type, std::string key, size_t val_size, bool random_value, uint32_t elements,
+    const CommandRegistry& registry, absl::InsecureBitGen* gen) {
+  absl::InlinedVector<string, 5> args;
+  args.push_back(std::move(key));
+
+  const CommandId* cid = nullptr;
+  if (type == "STRING") {
+    cid = registry.Find("SET");
+    args.push_back(GenerateValue(val_size, random_value, gen));
+  } else if (type == "LIST") {
+    cid = registry.Find("LPUSH");
+    for (uint32_t i = 0; i < elements; ++i) {
+      args.push_back(GenerateValue(val_size, random_value, gen));
+    }
+  } else if (type == "SET") {
+    cid = registry.Find("SADD");
+    for (size_t i = 0; i < elements; ++i) {
+      args.push_back(GenerateValue(val_size, random_value, gen));
+    }
+  } else if (type == "HSET") {
+    cid = registry.Find("HSET");
+    for (size_t i = 0; i < elements; ++i) {
+      args.push_back(GenerateValue(val_size / 2, random_value, gen));
+      args.push_back(GenerateValue(val_size / 2, random_value, gen));
+    }
+  } else if (type == "ZSET") {
+    cid = registry.Find("ZADD");
+    for (size_t i = 0; i < elements; ++i) {
+      args.push_back(absl::StrCat((*gen)() % val_size));
+      args.push_back(GenerateValue(val_size, random_value, gen));
+    }
+  } else if (type == "JSON") {
+    cid = registry.Find("JSON.SET");
+    args.push_back("$");
+
+    string json = "{";
+    for (size_t i = 0; i < elements; ++i) {
+      absl::StrAppend(&json, "\"", i, "\":\"", GenerateValue(val_size, random_value, gen), "\",");
+    }
+    json[json.size() - 1] = '}';  // Replace last ',' with '}'
+    args.push_back(json);
+  }
+  return {cid, args};
+}
+
+void DoPopulateBatch(string_view type, string_view prefix, size_t val_size, bool random_value,
+                     int32_t elements, const PopulateBatch& batch, ServerFamily* sf,
+                     ConnectionContext* cntx) {
+  boost::intrusive_ptr<Transaction> local_tx =
```

Comment:
I tried using the squasher API: https://github.com/dragonflydb/dragonfly/pull/2561/commits/9f3c5d357495e0e04f01f3e3f72112cdb61dc3df
but unfortunately it requires much more boilerplate (in the form of 3 vectors for different span conversions), and also doesn't work because current cntx doesn't have a tx. I imagine @dranikpg will know how to solve the second, but given the first I'm not sure if it's worth it..

Reply:
There is no point in it actually because this function is already is parallelized externally